File: ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/TransactionState.java
package net.corda.v5.ledger.utxo;

import net.corda.v5.base.annotations.CordaSerializable;
import net.corda.v5.base.annotations.DoNotImplement;
import net.corda.v5.ledger.common.Party;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Defines a transaction state, composed of a {@link ContractState} and associated transaction state information.
 *
 * @param <T> The underlying type of the {@link ContractState} instance.
 */
@DoNotImplement
@CordaSerializable
public interface TransactionState<T extends ContractState> {

    /**
     * Gets the {@link ContractState} of the current {@link TransactionState} instance.
     *
     * @return Returns the {@link ContractState} of the current {@link TransactionState} instance.
     */
    @NotNull
    T getContractState();

    /**
     * Gets the {@link ContractState} type of the current {@link TransactionState} instance.
     *
     * @return Returns the {@link ContractState} type of the current {@link TransactionState} instance.
     */
    @NotNull
    Class<T> getContractStateType();

    /**
     * Gets the {@link Contract} type of the current {@link TransactionState} instance.
     *
     * @return Returns the {@link Contract} type of the current {@link TransactionState} instance.
     */
    @NotNull
    Class<? extends Contract> getContractType();

    /**
     * Gets the notary of the current {@link TransactionState} instance.
     *
     * @return Returns the notary of the current {@link TransactionState} instance.
     */
    @NotNull
    Party getNotary();

    /**
     * Gets the encumbrance of the current {@link TransactionState} instance.
     *
     * @return Returns the encumbrance of the current {@link TransactionState} instance.
     */
    @Nullable
    EncumbranceGroup getEncumbranceGroup();
}

File: ledger/ledger-utxo/src/main/kotlin/net/corda/v5/ledger/utxo/TransactionState.kt
package net.corda.v5.ledger.utxo

import net.corda.v5.base.annotations.CordaSerializable
import net.corda.v5.base.annotations.DoNotImplement
import net.corda.v5.ledger.common.Party

/**
 * Defines a transaction state, composed of a [ContractState] and associated transaction state information.
 *
 * @param T The underlying type of the [ContractState] instance.
 * @property contractState The [ContractState] of the current [TransactionState] instance.
 * @property contractStateType The [ContractState] type of the current [TransactionState] instance.
 * @property contractType The [Contract] type of the current [TransactionState] instance.
 * @property notary The notary of the current [TransactionState] instance.
 * @property encumbrance The encumbrance of the current [TransactionState] instance.
 */
@DoNotImplement
@CordaSerializable
interface TransactionState<out T : ContractState> {
    val contractState: T
    val contractStateType: Class<out T>
    val contractType: Class<out Contract>
    val notary: Party
    val encumbrance: EncumbranceGroup?
}
